### DISCUSSION TITLE
Soroban Lottery Smart Contract

### DISCUSSION THREAD

**Discussion by heytdep** [stellar/sorobanathon#21] [General]
Hi there, I have also written a lottery smart contract and a description of how it works.

This is the repo: https://github.com/Xycloo/soroban-lottery-contract

## Contract Workflow
A lottery is a perfect example of showcasing the potential of smart contracts: transparency in the code and in its execution + ease of transferring assets. To create a lottery smart contract, we need it to do three things:
- Set up the lottery → the admin, which coin/token to accept as payment, how many winners there will be, how much will a lottery ticket cost (in terms of the accepted coin/token)
- Buying the ticket → swapping the required amount of the accepted token for a lottery ticket, thus a possiblity of winning.
- Running the lottery → When the admin wants (or when a certain condition is triggered), the lottery contract will randomly select and reward as many winners as the admin defined when setting up the lottery.

> In the current design, the lottery reward is made from the sales of the tickets (and the admin makes a small profit by receiving the remainder of the division of the prize between winners), but one could easily change the implementation to have a lottery with fixed prize and to send the ticket sales profit to the admin rather than keeping them in the contract.

### Fastrand
While for the vault contract I haven’t used other crates besides the soroban sdk and its auth helpers, you’ll notice that the Cargo.toml file for this contract uses fastrand, which the contract uses to generate random numbers for our winners. This is a good choice since we don’t need cryptographically secure numbers and this crate is light (0 dependencies).

```toml
[dependencies]
soroban-sdk = "0.1.0"
soroban-auth = "0.1.0"
fastrand = "1.8.0"
```

## Testing the contract
Since the contract relies on randomness, I can't make some assertions in the tests, so to see the results of the lottery after testing you can use the `-- --nocapture` flag:

```
> git clone https://github.com/Xycloo/soroban-lottery-contract
> cd soroban-lottery-contract
> cargo test -- --nocapture
   Compiling soroban-vault-contract v0.0.0 (/home/tommasodeponti/Desktop/soroban-lottery-contract)
    Finished test [unoptimized + debuginfo] target(s) in 2.19s
     Running unittests src/lib.rs (target/debug/deps/soroban_vault_contract-3af336ea264d8fce)

running 2 tests
test test::test ... ok


BALANCES
user1: 996
user2: 995
user3: 995 
user4: 995 
user5: 1017
user6: 1017 
user7: 995
user7: 995
user9: 995

test test::test_sequence ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.03s

   Doc-tests soroban-vault-contract

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

This submission was ideated and written along with @itsfederico.

**Comment by rice2000** [comment]
Fantastic!  Thanks for this.  More soon.

**Comment by achallagundla** [comment]
Heya Tommaso, 
Congratulations!  Your contribution to Sorobanathon: First Light has been accepted! 🎉
Your submission may qualify for a reward of 1000 XLM!  One last step: to receive that reward, please fill out this [form](https://stellarform.typeform.com/Sorobanathon) completely and accurately. 
 
Please let us know if you have any questions. (You can also message me on discord @ anuxhya)
 
All the best,
Anuhya

**Comment by leighmcculloch** [comment]
When I build the contract for wasm32 I get the following error:

```
$ cargo build --target wasm32-unknown-unknown --release
   Compiling soroban-vault-contract v0.0.0 (/workspace/Code/soroban-lottery-contract)
error: duplicate lang item in crate `std` (which `fastrand` depends on): `panic_impl`.
  |
  = note: the lang item is first defined in crate `soroban_sdk` (which `soroban_vault_contract` depends on)
  = note: first definition in `soroban_sdk` loaded from /workspace/Code/soroban-lottery-contract/target/wasm32-unknown-unknown/release/deps/soroban_sdk-84005e480b794e93.wasm, /workspace/Code/soroban-lottery-contract/target/wasm32-unknown-unknown/release/deps/libsoroban_sdk-84005e480b794e93.rlib
  = note: second definition in `std` loaded from /workspace/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/wasm32-unknown-unknown/lib/libstd-ef69f36ff320952a.rlib

error: could not compile `soroban-vault-contract` due to previous error
```

This is happening because `fastrand` is dependent on the Rust `std` package, which is pulling in components (the panic impl) that conflict with the panic impl that is provided by the Soroban SDK. The Soroban SDK provides a panic impl because it is intended to be used without `std`.

The `fastrand` impl is also not cryptographically secure.

@graydon Do you have any suggestions for how to produce randomness in contract at the moment? I know it was something you had discussed at some point for the future.

**Reply by heytdep** [reply]
@leighmcculloch I didn't actually think fastrand was dependent on the std, my bad for not checking. I had also tried using pointers and then the modulo to obtain "randomness" but that was obviously insecure as a number (just like using timestamps and then modulo would be). Open to any suggestions, in the meantime, I'll try and find some webassembly compatible crates, though I guess that something built directly in the host env would be very useful. 

That said, I don't know if there is a way to have genuinely cryptographically secure numbers without oracles. I had seen some posts about random number generation on the EVM, but the proposed solution was unfit (at least IMO) for an example first-flight contract (i.e too complex and not really scalable as there were many constraints).

**Reply by graydon** [reply]
I've put together a preliminary PRNG interface for the soroban host in https://github.com/stellar/rs-soroban-env/pull/544 which, while imperfect, should be as unpredictable as the existing transaction-ordering mechanism in stellar-core, and can be strengthened over time without changing the contract interface. Comments welcome; if it seems appropriate we'll try to get this into an upcoming release.

**Comment by heytdep** [comment]
@leighmcculloch @graydon I have cloned https://github.com/stellar/rs-soroban-env/pull/544 and tinkered a bit with the PRNG interface in a contract by adding a simple helper to the SDK. Seems to be working just fine:

![prng_rs_tests_](https://user-images.githubusercontent.com/70587974/197007977-7f9abbfb-408f-458c-8c83-30c518cfa14a.png)

PS: ignore the fact that the prng method is at `ledger()`, probably not a good place for such function 😅

**Reply by leighmcculloch** [reply]
Very cool!!! 👏🏻
